### ISSUE TITLE
Correct way to handle attachments?

### ISSUE THREAD

**Issue by Tongboy** [minimul/qbo_api#42]
Our app needs to manipulate attachable objects, doesn't look like support for file upload exists today.

Before I put together a pull request I wanted to see if there was any thoughts as to how to implement it.

The main difference is the faraday request needs to be multipart form rather than a json request and the json request that goes along with the file goes as a separate file attachment rather than as JSON in the body. The paths for file operations are also different than the standard REST endpoints. It's quite a bit different than every other request/response from the server for other types.

I figure it's prudent to fit it in to the existing pattern as much as possible but I could also see using different methods to work with it - is there any preference?

I put together a quick module hack that successfully uploads files and generally fits in the same `api.method(:resource, payload: payload)` pattern with the exception of needing to pass in a `file` hash param for the file to be uploaded.
```ruby
module QboAttachments
  def connection(url: get_endpoint, multipart: false)
    Faraday.new(url: url) do |faraday|
      faraday.headers['Content-Type'] = 'application/json;charset=UTF-8' unless multipart
      faraday.headers['Accept'] = 'application/json'
      puts 'SETTING MULTIPART' if multipart
      if !@token.nil?
        faraday.request :oauth, oauth_data
      elsif !@access_token.nil?
        faraday.request :oauth2, @access_token, token_type: 'bearer'
      else
        raise QboApi::Error.new error_body: 'Must set either the token or access_token'
      end
      faraday.request :multipart if multipart
      faraday.request :url_encoded
      faraday.use FaradayMiddleware::RaiseHttpException
      faraday.response :detailed_logger, QboApi.logger if QboApi.log
      faraday.adapter  Faraday.default_adapter
    end
  end

  def request(method, path:, entity: nil, payload: nil, params: nil)
    multipart = entity == :attachable
    raw_response = connection(multipart: multipart).send(method) do |req|
      path = finalize_path(path, method: method, params: params)
      case method
      when :get, :delete
        req.url path
      when :post, :put
        req.url path
        if multipart
          req.url finalize_path(path, method: method, params: params).gsub('attachable', 'upload')
          req.body = { file_metadata_01: Faraday::UploadIO.new(StringIO.new(JSON.generate(payload.except('file'))), 'application/json', 'attachment.json'),
                       file_content_01: payload['file'] }
        else
          req.body JSON.generate(payload)
        end
      end
    end
    response(raw_response, entity: entity)
  end
end
QboApi.prepend(QboAttachments)

> api = QboApi.new(...)
> payload = { AttachableRef: [{ EntityRef: {type: 'Invoice', value: '9'}, IncludeOnSend: 'true' }], FileName: 'test.txt', ContentType: 'text/plain', file: Faraday::UploadIO.new('test.txt', 'text/plain') }
> api.create(:attachable, payload: payload)
(returns an AttachableResponse and invoice 9 now has a visible attachment of test.txt on it)
```

Thank you

**Comment by minimul** [comment]
Strangely, I have not had this particular need yet but will absolutely get a PR around this in.

@ntippie Can you advise when you have a moment?

**Comment by ntippie** [comment]
This implementation conflicts with the PR I submitted earlier today, which stores the connection as an instance variable (see #43).

I made another implementation that you can see [on my fork](https://github.com/minimul/qbo_api/compare/master...ntippie:attachments). I wanted to get more feedback from you both before submitting a PR.

Specific things I'd like feedback on:

- This creates a new `#attachment_connection` method, which somewhat conflicts with #43 in that you have to edit the middleware for each connection. Perhaps it would be better for `#attachment_connection` to dup `#connection` and add the multipart middleware and correct Content-Type header.
- This uses a new method, `#upload_attachment`, instead of `#create`. I think this is more in line with the current level of API abstraction (ie, special methods for special endpoints).
- Like @Tongboy's proposed implementation, this only supports uploading a single attachment at a time. [The API docs](https://developer.intuit.com/docs/api/accounting) specify that it is possible to upload many at once, as long as the request size is less than 100MB.
- Errors are returned as seen below. Unfortunately we cannot use `raise_http_exception.rb` for this, since such a response has a status code of 200 as it is possible to have mixed success/failure if uploading multiple attachments in the same request.

```ruby
{"AttachableResponse"=>
  [{"Fault"=>
     {"Error"=>
       [{"Message"=>"Object Not Found",
         "Detail"=>
          "Object Not Found : Something you're trying to use has been made inactive. Check the fields with accounts, customers, items, vendors or employees.",
         "code"=>"610",
         "element"=>""}],
      "type"=>"ValidationFault"}}],
 "time"=>"2018-01-03T13:06:31.406-08:00"}
```

**Comment by Tongboy** [comment]
That's a lot cleaner than trying to bludgeon it all in to the existing methods!

Add a `#delete_attachment` or similar and you'll have the two common interactions covered for attachments.

You could add multiple attachment support without *too* much pain to your `#upload_attachment` method with some arrays/`Array()` but you run in to a lot more checking to make sure things match. Seems far easier to leave it singular for now until the need exists where multiple single attachments is a limiter for some reason.

**Comment by ntippie** [comment]
You should be able to use the `#delete` method to delete an attachment, I believe.

Perhaps we should consider the multiple attachment support before merging. Something like your `payload['file']` might work better since you'd need to have the attachment matched with the payload.

I also considered checking for `response['AttachableResponse'].first['Fault']` to construct and raise a `QboApi::BadRequest` error, but we should probably decide on the multiple attachments first.

More testing is probably needed before implementing multiple attachments. I'm not sure what happens if you breach the 100MB limit, or what happens if you have more than 99 attachments (ie, if `file_content_100` is a valid multipart title).

**Comment by Tongboy** [comment]
Fair point on the `#delete` - from an outsider's perspective I'd think it would be helpful to have a 'helper' method that falls inline with the `#upload_attachment` if just for simplicity/parity.

I'll test the bahavior on the `file_content_100` and upload file size right now

**Comment by ntippie** [comment]
Since this endpoint seems to function most similarly to `#batch`, I think I was wrong in suggesting that we raise an error if there is a fault in the request. However, I'm still undecided on how best to pair the file_metadata and file_content for the `#upload_attachment` (`#upload_attachments`?) method. Embed the file in the payload? Use a hash or 2D array? I'm not sure-- what do you guys think?

**Comment by minimul** [comment]
Agreed that @ntippie 's diff looks good.

I vote for single attachment support for this iteration but will merge a multi-attachment PR.

The `response['AttachableResponse'].first['Fault']` is useful for the library to handle and raise but stinks that Intuit is returning a `200` on those. Therefore, it would be OK to leave that off the PR as well and in its place a README note that you'll need to sniff for an attachment error.

**Comment by ntippie** [comment]
@minimul Would you like me to make a PR from that branch, then? Any other edits, like version numbers? I suppose we can refactor later on and add a `#upload_attachments`, but with the current `#upload_attachment` arguments, the API between the single/multiple methods would be quite different.

**Comment by Tongboy** [comment]
File size over 100 (which took substantially increasing request timeouts even on really fast internet) returns a 400 bad request `{"message":"Bad Request","errorCode":"400","cause":"CLIENT"}"`

**Comment by ntippie** [comment]
I saw the same exact error when I failed to specify a content-type and filename for the file-content, and it caused an error in `RaiseHttpException#parse_json` (not a designed error) because it does not follow the usual format.

**Comment by minimul** [comment]
@Tongboy Thanks for checking on that – at least that error will get raised.

@ntippie Yes, please submit that PR (when you have time – of course). I can contribute by adding the spec(s).

@Tongboy I'll then get this enhancement released at version 1.6.0.

**Comment by ntippie** [comment]
Just submitted the PR. I think the current arguments for `#upload_attachment` are good, and that the method for multiple attachments should use a 2D array.

**Comment by minimul** [comment]
Thanks @ntippie 

Released => 1.6.0

**Comment by minimul** [comment]
FYI: 1.6.1 => [PR](https://github.com/minimul/qbo_api/pull/48)
